index/segment/fst/writer.go

```diff
@@ -0,0 +1,361 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fst
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/generated/proto/fstwriter"
+	"github.com/m3db/m3ninx/index/segment"
+	"github.com/m3db/m3ninx/index/segment/fs/encoding"
+	"github.com/m3db/m3ninx/postings/pilosa"
+)
+
+var (
+	defaultInitialPostingsOffsetsMapSize = 1024
+	defaultInitialFSTTermsOffsetsMapSize = 1024
+	defaultInitialIntEncoderSize         = 128
+
+	errUnableToFindPostingsOffset = errors.New("internal error: unable to find postings offset")
+	errUnableToFindFSTTermsOffset = errors.New("internal error: unable to find fst terms offset")
+)
+
+type writer struct {
+	seg segment.Segment
+
+	intEncoder      *encoding.Encoder
+	postingsEncoder *pilosa.Encoder
+	fstWriter       *fstWriter
+
+	metadata            []byte
+	docsDataFileWritten bool
+	postingsFileWritten bool
+	fstTermsFileWritten bool
+	postingsOffsets     *postingsOffsetsMap
+	fstTermsOffsets     *fstTermsOffsetsMap
+}
+
+// NewWriter returns a new writer.
+func NewWriter() Writer {
+	return &writer{
+		intEncoder:      encoding.NewEncoder(defaultInitialIntEncoderSize),
+		postingsEncoder: pilosa.NewEncoder(),
+		fstWriter:       newFSTWriter(),
+		postingsOffsets: newPostingsOffsetsMap(defaultInitialPostingsOffsetsMapSize),
+		fstTermsOffsets: newFSTTermsOffsetsMap(defaultInitialFSTTermsOffsetsMapSize),
+	}
+}
+
+func (w *writer) Clear() {
+	w.seg = nil
+
+	w.intEncoder.Reset()
+	w.postingsEncoder.Reset()
+	w.fstWriter.Clear()
+
+	w.metadata = nil
+	w.docsDataFileWritten = false
+	w.postingsFileWritten = false
+	w.fstTermsFileWritten = false
+	w.postingsOffsets.Reset()
+	w.fstTermsOffsets.Reset()
+}
+
+func (w *writer) Reset(s segment.MutableSegment) error {
+	w.Clear()
+
+	numDocs := s.Size()
+	metadata := defaultV1Metadata()
+	metadata.NumDocs = numDocs
+	metadataBytes, err := metadata.Marshal()
+	if err != nil {
+		return err
+	}
+
+	w.metadata = metadataBytes
+	w.seg = s
+	return nil
+}
+
+func (w *writer) MajorVersion() int {
+	return majorVersion
+}
+
+func (w *writer) MinorVersion() int {
+	return minorVersion
+}
+
+func (w *writer) Metadata() []byte {
+	return w.metadata
+}
+
+func (w *writer) WriteDocumentsData(iow io.Writer) error {
+	return nil
+}
+
+func (w *writer) WriteDocumentsIndex(iow io.Writer) error {
+	if !w.docsDataFileWritten {
+		return fmt.Errorf("documents data file has to be written before documents index file")
+	}
+	return nil
+}
+
+func (w *writer) WritePostingsOffsets(iow io.Writer) error {
+	currentOffset := uint64(0)
```

Comment:
Shouldn't this also be `if !w.docsDataFileWritten || !w.docsDataIndexFileWritten { ... }`?  Can stub out in tests by forcing them to be true for now I guess if it was a pain to test them out.

Reply:
Nah, so the identifiers in the postings data is `doc.ID`, which doesn't depend on where/how the documents data is persisted. So don't need the ordering constraint here.